tests/garage/tf/core/test_gru.py

```diff
@@ -0,0 +1,390 @@
+
+        self.step_hidden_var = tf.placeholder(
+            shape=(self.batch_size, self.hidden_dim),
+            name='initial_hidden',
+            dtype=tf.float32)
+
+        self.gru_cell = tf.keras.layers.GRUCell(
+            units=self.hidden_dim,
+            activation=tf.nn.tanh,
+            kernel_initializer=tf.constant_initializer(1),
+            recurrent_activation=tf.nn.sigmoid,
+            recurrent_initializer=tf.constant_initializer(1),
+            name='lstm_layer')
+
+    # yapf: disable
+    @params(
+        (1, 1, 1, 0, 0),
+        (1, 1, 3, 0, 0),
+        (1, 3, 1, 0, 0),
+        (3, 1, 1, 0, 0),
+        (3, 3, 1, 0, 0),
+        (3, 3, 3, 0, 0),
+        (1, 1, 1, 0.5, 0.5),
+        (1, 1, 3, 0.5, 0.5),
+        (1, 3, 1, 0.5, 0.5),
+        (3, 1, 1, 0.5, 0.5),
+        (3, 3, 1, 0.5, 0.5),
+        (3, 3, 3, 0.5, 0.5))
+    # yapf: enable
+    def test_output_shapes(self, time_step, input_dim, output_dim, hidden_init,
+                           cell_init):
+        obs_inputs = np.full((self.batch_size, time_step, input_dim), 1.)
+        obs_input = np.full((self.batch_size, input_dim), 1.)
+
+        input_var = tf.placeholder(
+            tf.float32, shape=(None, None, input_dim), name='input')
+        step_input_var = tf.placeholder(
+            tf.float32, shape=(None, input_dim), name='input')
+        output_nonlinearity = tf.keras.layers.Dense(
+            units=output_dim,
+            activation=None,
+            kernel_initializer=tf.constant_initializer(1))
+        with tf.variable_scope('GRU'):
+            self.gru = gru(
+                all_input_var=input_var,
+                name='gru',
+                gru_cell=self.gru_cell,
+                step_input_var=step_input_var,
+                step_hidden_var=self.step_hidden_var,
+                hidden_state_init=tf.constant_initializer(hidden_init),
+                output_nonlinearity_layer=output_nonlinearity)
+
+        self.sess.run(tf.global_variables_initializer())
+
+        # Compute output by doing t step() on the gru cell
+        outputs_t, output_t, h_t, hidden_init = self.gru
+        hidden = np.full((self.batch_size, self.hidden_dim),
+                         hidden_init.eval())
+
+        for i in range(time_step):
+            output, hidden = self.sess.run([output_t, h_t],
+                                           feed_dict={
+                                               step_input_var: obs_input,
+                                               self.step_hidden_var: hidden,
+                                           })  # noqa: E126
+            assert output.shape == (self.batch_size, output_dim)
+            assert hidden.shape == (self.batch_size, self.hidden_dim)
+
+        full_output = self.sess.run(
+            outputs_t, feed_dict={input_var: obs_inputs})
+
+        assert full_output.shape == (self.batch_size, time_step, output_dim)
+
+    # yapf: disable
+    @params(
```

Comment:
Could you add a test of variable length inputs to GRU and LSTM? It seems we don't have any tests using off-policy algorithms with recurrent policies. And since DQN doesn't set a max_path_length, I assume the episodes would have variable lengths. For recurrent policies, we should pad them. Do we pad it somewhere? Please point it to me if we do.

Reply:
If we pad the inputs, then we don't have to worry about variable length inputs?

Reply:
Sorry just saw this. 
I think I might have used an incorrect example. It is uncommon to use LSTMNetwork in DQN. Besides, during training, the shape of samples (from replay buffer) would be `(sample_size, 1, obs_dim)`. During sampling, the shape would be `(1, time_steps, obs_dim)`. So it might not be a problem in the current off-policy algorithms. 
If we pad them, we just need to indicate that `<pad>` is a condition to stop the traverse. `tf.while_loop` or `tf.scan`(what we are using) should have an argument to pass the condition (please check, I am not familiar with `tf.scan`).

Reply:
I think this problem seems to beyond the pr's scope. If you think we should address this later or we don't have the situation to deal with variable-length inputs, feel free to ignore this issue.